### ISSUE TITLE
Moving Transit Lines to Managed Lane

### ISSUE THREAD

**Issue by e-lo** [wsp-sag/network_wrangler#51]
Create a method in the TransitNetwork to move transit lines, as specified, to a parallel managed lane facility.

This may include querying which transit lines have stops that can be served if they are moved vs not.

**Comment by josiekre** [comment]
Can you provide an example project card? Mainly this is so that I can get my head around how this will come from Cube and therefore via the project card. 

See #61 for the latest example project card on transit.

**Comment by e-lo** [comment]
basically it should look like a managed lane project card...but
I added a new keyword: `actions`

Open to other ideas


```
project: test managed lane project with transit moving
category: Parallel Managed Lanes
tags:
   - 'example'
facility:
  link:
    - name:
      - 'I 35E'
  A:
    osmNodeId: '961117623'
  B:
    osmNodeId: '2564047368'
actions:
 - move_transit:
      - time: ['7:00', '9:00']
      - short_name: ...
properties:
  - property: LANES
    set: 3
    timeofday:
      - time: ['7:00', '9:00']
        set: 2
      - time: ['16:00', '19:00']
        set: 2
  - property: ML_LANES
    set: 0
    timeofday:
      - time: ['7:00', '9:00']
        set: 1
      - time: ['16:00', '19:00']
        set: 1
  - property: ML_PRICE
    set: 0
    group:
      - category: ['sov']
        timeofday:
          - time: ['7:00', '9:00']
            set: 1.5
          - time: ['16:00', '19:00']
            set: 2.5
      - category: ['hov2']
        timeofday:
          - time: ['7:00', '9:00']
            set: 1.0
          - time: ['16:00', '19:00']
            set: 2.0
  - property: ML_ACCESS
    set: 'all'
  - property: ML_EGRESS
    set: 'all'
notes: Placeholder for notes
```

**Comment by josiekre** [comment]
So many questions immediately come to mind...


Where does the graph information come from? How does TransitNetwork know what node IDs to move (existing, set)?

Why not have the project card use a "Transit Service Property Change" type to specify this instead? In the case project cards allow multiple types of changes in one card, Lasso would already have the node list and TransitNetwork could handle it like any other routing change.

**Comment by e-lo** [comment]
Well one of these is at least easier...

> Where does the graph information come from?

The link list, node list (route), and graph already exist from the selection query done in the roadway network. It can be accessed by using the following methods:

1. searches are stored by selection key, so create a selection key
https://github.com/wsp-sag/network_wrangler/blob/0f06a39b66d1f34d7a71f96113f7c1b1c952f178/network_wrangler/RoadwayNetwork.py#L454

2. https://github.com/wsp-sag/network_wrangler/blob/0f06a39b66d1f34d7a71f96113f7c1b1c952f178/network_wrangler/RoadwayNetwork.py#L628-L632

**Comment by e-lo** [comment]
> Why not have the project card use a "Transit Service Property Change" type to specify this instead? 

I think you mean route change?

> Lasso would already have the node list

I'm not sure how Lasso would already have the node list? Wrangler is treating each card individually in apply. In any case, we don't want to rely on Lasso for this functionality...it should be a core wrangler functionality.

**Comment by josiekre** [comment]
> I think you mean route change?

I put them into the same project card type. Is this a problem? Right now we only handle `headway_secs` and `routing` property changes so it seemed cleaner to use the existing functions rather than creating another project card type with a `routing` key instead of a `properties` key. 

```
project: Bus Reroute
category: Transit Service Property Change
tags:
   - existing_plus_committed
dependencies:
  prerequisites: []
  corequisites: []
  conflicts: []
facility:
  trip_id: 14944022-JUN19-MVS-BUS-Weekday-01
properties:
  - property: routing
    existing: [-1930, -1931, 1932]
    set: [-1930, 1931, 1932]
```

**Comment by josiekre** [comment]
- [ ] Add class "global" MANAGED_LANES_SCALAR = 10000 to RoadwayNetwork
- [ ] Create get_managed_lane_nodes() in RoadwayNetwork (adds scalar to existing nodes)
- [ ] Implement apply_feature_change_managedlane() inside TransitNetwork, which will call get_managed_lane_nodes to implement routing change
- [ ] Implement select_trip_id_by_nodes() inside TransitNetwork #61

**Comment by e-lo** [comment]
@josiekre  - just chatted with @kulshresthaa  and saw that the scalar for nodes will need to be 500,000

**Comment by josiekre** [comment]
- [ ] Query node_ids to identify stops (negative/positive indicators from project cards will not be present when pulling node_ids from RoadwayNetwork.selection.route)

**Comment by josiekre** [comment]
@e-lo @kulshresthaa Do you have a project card (or other test) you have created already that would have transit service that needs to move? It will otherwise take me a bit of time to think up data for a test. Running out of time available so trying to save where I can.

**Comment by e-lo** [comment]
we don't. but I'll make one and you can change as you need to

**Comment by josiekre** [comment]
I think the new St Paul network needs to be brought over to test this cause RoadwayNetwork's graph implementation is possibly using a different ID (OSM?).

**Comment by e-lo** [comment]
you can pull my branch which passes tests as of now and is ready to be
merged to develop

**Comment by josiekre** [comment]
Working on it. Merge into my branch is a bit of a doozy. Lots of conflicts.

**Comment by josiekre** [comment]
Not so bad. Only thing left is `stops.txt` and `shapes.txt`. Did you pull in those updates from Lasso with node IDs instead of link IDs? If so, I'll merge yours over mine.

**Comment by josiekre** [comment]
I think I resolved the above. I'm guessing you used a Find and Replace on variable names, so I stuck with osm_node_id over osm_link_id (which was the change from your branch).

Any who, after the merge the roadway network example files are the same on my branch as on yours but I'm getting a failure when I try to read it in. With fast=False, schema failure. With fast=True, type error.

```
    @staticmethod
    def read(
        link_file: str, node_file: str, shape_file: str, fast: bool = False
    ) -> RoadwayNetwork:
        """
        Reads a network from the roadway network standard
        Validates that it conforms to the schema
    
        args:
        link_file: full path to the link file
        node_file: full path to the node file
        shape_file: full path to the shape file
        fast: boolean that will skip validation to speed up read time
        """
        if not fast:
            if not (
                RoadwayNetwork.validate_node_schema(node_file)
                and RoadwayNetwork.validate_link_schema(link_file)
                and RoadwayNetwork.validate_shape_schema(shape_file)
            ):
    
                sys.exit("RoadwayNetwork: Data doesn't conform to schema")
    
        with open(link_file) as f:
            link_json = json.load(f)
    
>       link_properties = pd.DataFrame(link_json["features"])
E       TypeError: list indices must be integers or slices, not str
```

**Comment by josiekre** [comment]
Did my best. Throwing in the towel for this evening. 

Besides testing this last feature (`apply_transit_managed_lane()`) and potentially dealing with different IDs between transit and RoadwayNetwork's graph, I think issues assigned to me for Sprint 7 are completed.

Unfortunately, I'm not available to work on this tomorrow. I may be able to over the weekend if no one can address it before then.

**Comment by e-lo** [comment]
@josiekre  - i just pushed a project card which you should update (when you get a chance) to be the format that you are expecting...but the node IDs should at least work with the graph.

Did you write a test?  I didn't see one but didn't know if I wasn't looking in the right place.  If not, please prioritize this so that we can evaluate if it is working.

**Comment by josiekre** [comment]
I have a few tests already for a routing change. I do not have a test for a managed lane project that also changes transit routing.
